# Expression heat maps

expr_heat_map_text <- 
"For the bulk RNA-seq analysis, Tregs and Tconv were isolated from blood, and perfused spleen, lymph nodes, kidney, liver, lung, pancreas, and gut (Peyer’s patches, LPLs, IELs) samples. Samples were prepared from Foxp3Thy1.1 reporter mice, allowing purification based on Foxp3 expression. mRNA was sequencing using the QuantSeq 3’mRNA-Seq Library Prep Kit for Illumina and the QuantSeg data analysis workflow. Heatmaps showing the expression of genes across all samples are based on either the 50 most expressed genes (per tissue) or the (maximum 50) most differentially-expressed genes (per tissue, in comparison to the blood Treg population). Select either the most expressed gene or most differentially-expressed gene option, and then select the tissue of choice, to show the heatmap for all samples based on the selected tissue. "

#Plots may be downloaded by right clicking the plot and selecting 'Save image as...' 

#Effect of genetic perturbation on tissue Treg phenotype

genetic_perturbation_text <- "High-dimensional flow cytometry phenotyping was performed on tissue Tregs with genetic deficiency in CD69, ST2, KLRG1, Itgae (CD103), CD11a, amphiregulin (Areg), S1PR2, Hif1α, Blimp1 and BATF. The data for CD69, ST2, KLRG1, CD11a, S1PR2, Hif1α, Blimp1 and BATF are presented from 50%:50% mixed bone-marrow chimeras, where wildtype-derived Tregs serve as the internal control for KO-derived Tregs. The data for Areg and CD103 are presented as analysis of KO mice compared to wildtype littermates. Data can be displayed as either cellular phenotypes or marker expression. Cellular phenotype will allow the selection of the genetic perturbation and the tissue, displaying a tSNE-based quantification of the Treg subsets present in wildtype and KO cells. Marker expression will allow the selection of the marker of choice, with expression displayed as a heatmap and as a tSNE expression overlay, across the different tissues, in wildtype and KO Tregs."
